enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/AbstractBanDependencies.java

```diff
@@ -18,22 +18,15 @@
  */
```

Comment:
I consider this file API (other rules might extend that) so it is backwards-incompatible to remove any non-private methods in there or change their signature. We may want to deprecate things though

Reply:
Really??? If so, I consider this a bad API. Especially because a collection of artifacts is unsuitable for a general purpose.

I might as well simply move the rules out of this class.

Reply:
API here just means it is exported via the class path to all custom rules and can be accessed and extended. It doesn't say anything about design....

Potentially that affects every rule which is not a final class (because there is no limited visibility) but I wouldn't go that far to say that it is strictly necessary to keep backwards compatibility but for abstract public classes I would say yes, we unfortunately need to be backwards compatible here (so first deprecate and only finally remove in m-enforcer-p 4.0.0)

Reply:
See also https://github.com/apache/maven-enforcer/pull/167.

Reply:
@slawekjaranowski Which classes do you consider so relevant that we need to ensure backwards-compatibility (again all of them are exposed to custom rules unfortunately)

Reply:
Only classes from `enforcer-api` should be used to write own rules, like we have in docs:
https://maven.apache.org/enforcer/enforcer-api/writing-a-custom-rule.html

even more in `extra-enforcer-rules` internal `enforcer-rules` are baned:
https://github.com/mojohaus/extra-enforcer-rules/blob/master/pom.xml#L147-L152

Reply:
I don’t think this is explicit enough in the docs. I know of several custom rules which leverage or extend enforcer rules classes.

Reply:
Imho, the documentation is clear on how to create custom rules -- enforcer-api must be supported. Is non-standard use supported? What if there are projects that depend on the current internal implementation? Should this also be supported?

Reply:
I must agree with Andrzej if someone use internal implementation, we should not stop our improvements in such case.

Documentation say that api should be used nothing else.

Yes we should think to hide internal classes when we can, but sometime it is not possible for technical reason.

Reply:
@kwin I checked list from mvnrepository ... only few of project was released in this year. Another are very old ... 
I think we can take a such risik and go forward. 

By the way documentation should be more clear in this point - TODO in next PR.